src/NServiceBus.Core/Recoverability/FirstLevelRetries/FailureInfoStorage.cs

```diff
@@ -0,0 +1,88 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+
+    // The data structure has fixed maximum size. When the data structure reaches its maximum size,
+    // the least recently used (LRU) message processing failure is removed from the storage.
+    class FailureInfoStorage
+    {
+        public FailureInfoStorage(int maxElements = 1000)
+        {
+            this.maxElements = maxElements;
+        }
+
+        public void RecordFailureInfoForMessage(string messageId, Exception exception)
+        {
+            lock (lockObject)
```

Comment:
@tmasternak We don't know but I have a hunch that this data structure will not experience lock contention for the following reasons:

- Not that many threads will access it,
- It will not be accessed in a tight loop by any of the threads. The threads will typically be involved in some I/O (timeout persistence or the transport) and when they fail, they will update the data structure.
- The `lock` construct is a hybrid lock and will do a little spin lock first and only then it will fall back to a OS synchronization primitive. What this means is that most likely spin locks will handle synchronization, which should be pretty efficient.

@danielmarbach I read that same book :).